redis/src/aio/connection.rs

```diff
@@ -256,76 +258,271 @@ where
+    /// Splits the [`PubSub`] into a [`SplitPubSubSink`] and [`SplitPubSubStream`].
+    pub fn split(self) -> (SplitPubSubSink<C>, SplitPubSubStream<C>) {
+        SplitPubSubSink::new(self.framed, self.queue)
+    }
+
+    /// Query the connection with a command and retrieves the response.
+    pub async fn query<'a, T>(&'a mut self, cmd: &'a Cmd) -> RedisResult<T>
+    where
+        T: FromRedisValue,
+    {
+        let mut out = Vec::new();
+
+        cmd.write_packed_command(&mut out);
+        self.framed.send(out).await?;
+
+        self.next_response().await
+    }
+
+    async fn next_response<T>(&mut self) -> RedisResult<T>
+    where
+        T: FromRedisValue,
+    {
+        while let Some(item) = self.framed.next().await {
+            match item {
+                Ok(Ok(value)) => match Msg::from_value(&value) {
+                    Some(msg) => self.queue.push_front(msg),
+                    _ => return from_redis_value(&value),
+                },
+                _ => break,
+            }
+        }
+
+        Err(stream_dropped_error())
     }
 
     /// Subscribes to a new channel.
     pub async fn subscribe<T: ToRedisArgs>(&mut self, channel: T) -> RedisResult<()> {
-        cmd("SUBSCRIBE").arg(channel).query_async(&mut self.0).await
+        self.query(cmd("SUBSCRIBE").arg(channel)).await
     }
 
     /// Subscribes to a new channel with a pattern.
     pub async fn psubscribe<T: ToRedisArgs>(&mut self, pchannel: T) -> RedisResult<()> {
-        cmd("PSUBSCRIBE")
-            .arg(pchannel)
-            .query_async(&mut self.0)
-            .await
+        self.query(cmd("PSUBSCRIBE").arg(pchannel)).await
     }
 
     /// Unsubscribes from a channel.
     pub async fn unsubscribe<T: ToRedisArgs>(&mut self, channel: T) -> RedisResult<()> {
-        cmd("UNSUBSCRIBE")
-            .arg(channel)
-            .query_async(&mut self.0)
-            .await
+        self.query(cmd("UNSUBSCRIBE").arg(channel)).await
     }
 
     /// Unsubscribes from a channel with a pattern.
     pub async fn punsubscribe<T: ToRedisArgs>(&mut self, pchannel: T) -> RedisResult<()> {
-        cmd("PUNSUBSCRIBE")
-            .arg(pchannel)
-            .query_async(&mut self.0)
+        self.query(cmd("PUNSUBSCRIBE").arg(pchannel)).await
+    }
+}
+
+impl Stream for PubSub {
+    type Item = Msg;
+
+    fn poll_next(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let PubSubProj { mut framed, queue } = self.project();
+
+        if let Some(msg) = queue.pop_back() {
+            return Poll::Ready(Some(msg));
+        }
+
+        loop {
+            match framed.as_mut().poll_next(cx) {
+                Poll::Ready(Some(Ok(Ok(value)))) => {
+                    if let Some(msg) = Msg::from_value(&value) {
+                        return Poll::Ready(Some(msg));
+                    }
+                }
+                Poll::Ready(Some(Err(_)) | Some(Ok(Err(_)))) => {}
```

Comment:
making sure - will the stream terminate once the pubsub is disconnected from the server? can we add tests that drop the server and see that the stream completes?